nvflare/client/api.py

```diff
@@ -39,36 +39,46 @@ def init(config: Union[str, Dict] = f"config/{CONFIG_EXCHANGE}", rank: Optional[
     """
     global PROCESS_MODEL_REGISTRY  # Declare PROCESS_MODEL_REGISTRY as global
 
-    if rank is None:
-        rank = os.environ.get("RANK", "0")
-
-    if PROCESS_MODEL_REGISTRY:
-        raise RuntimeError("Can't call init twice.")
-
-    if isinstance(config, str):
-        client_config = from_file(config_file=config)
-    elif isinstance(config, dict):
-        client_config = ClientConfig(config=config)
-    else:
-        raise ValueError("config should be either a string or dictionary.")
-
-    dx = None
-    if rank == "0":
-        if client_config.get_exchange_format() == ExchangeFormat.PYTORCH:
-            tensor_decomposer, ok = optional_import(module="nvflare.app_opt.pt.decomposers", name="TensorDecomposer")
-            if ok:
-                fobs.register(tensor_decomposer)
+    try:
+        if rank is None:
+            rank = os.environ.get("RANK", "0")
+
+        if PROCESS_MODEL_REGISTRY:
+            raise RuntimeError("Can't call init twice.")
+
+        if isinstance(config, str):
+            client_config = from_file(config_file=config)
+        elif isinstance(config, dict):
+            client_config = ClientConfig(config=config)
+        else:
+            raise ValueError("config should be either a string or dictionary.")
+
+        dx = None
+        if rank == "0":
+            if client_config.get_exchange_format() == ExchangeFormat.PYTORCH:
+                tensor_decomposer, ok = optional_import(
+                    module="nvflare.app_opt.pt.decomposers", name="TensorDecomposer"
+                )
+                if ok:
+                    fobs.register(tensor_decomposer)
+                else:
+                    raise RuntimeError(f"Can't import TensorDecomposer for format: {ExchangeFormat.PYTORCH}")
+
+            pipe_args = client_config.get_pipe_args()
+            if client_config.get_pipe_class() == "FilePipe":
+                pipe = FilePipe(**pipe_args)
             else:
-                raise RuntimeError(f"Can't import TensorDecomposer for format: {ExchangeFormat.PYTORCH}")
+                raise RuntimeError(f"Pipe class {client_config.get_pipe_class()} is not supported.")
 
-        # TODO: make Pipe configurable
-        dx = FilePipeDataExchanger(
-            supported_topics=client_config.get_supported_topics(),
-            data_exchange_path=client_config.get_exchange_path(),
-            pipe_name=client_config.get_pipe_name(),
-        )
+            dx = DataExchanger(
+                supported_topics=client_config.get_supported_topics(),
+                pipe=pipe,
+                pipe_name=client_config.get_pipe_name(),
+            )
```

Comment:
Does it miss an "else" condition here?

Reply:
@yhwen no, only rank 0 is going to communicate with NVFlare side.